sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/NumberPatternShouldBeRegular.cs

```diff
@@ -57,42 +58,97 @@ protected override void Initialize(SonarAnalysisContext context)
 
                     var literal = (LiteralExpressionSyntax)c.Node;
 
-                    var numberWithoutSuffix = ClearNumberTypeSuffix(literal.Token.Text);
-
-                    var decimalParts = numberWithoutSuffix.Split('.');
-                    if (decimalParts.Length > 2)
-                    {
-                        return;
-                    }
-
-                    var hasIrregularPattern = decimalParts
-                        .SelectMany(part => part.Split('_'))
-                        .Select(x => x.Length)
-                        .Skip(1) // skip the first part (1_234 => 234)
-                        .Reverse().Skip(decimalParts.Length == 2 ? 1 : 0) // skip the last if there is a decimal (.234_5 => 234)
-                        .Distinct().Skip(1).Any(); // we expect to have only 1 size of pattern
-
-                    if (hasIrregularPattern)
+                    if (HasIrregularPattern(literal.Token.Text))
                     {
                         c.ReportDiagnosticWhenActive(Diagnostic.Create(rule, literal.GetLocation()));
                     }
                 },
                 SyntaxKind.NumericLiteralExpression);
         }
 
-        private static string ClearNumberTypeSuffix(string numberLiteral)
+        /// <remarks>internal for test purposes.</remarks>
+        internal static bool HasIrregularPattern(string numericToken)
+        {
+            var split = StripNumericPreAndSuffix(numericToken).Split(Dot);
+
+            // ignore multiple dots.
+            if (split.Length > 2)
+            {
+                return false;
+            }
+
+            var groups = split[0].Split(Underscore).Select(s => s.Length).ToArray();
+
+            var size = NotFound;
+
+            if (groups.Length > 1)
+            {
+                size = groups[1];
+
+                // first should not be bigger.
+                if (groups[0] > size)
+                {
+                    return true;
+                }
+
+                // all should have the same size, except the first, and the size should at least be 2.
+                if (size < 2 || groups.Skip(1).Any(g => g != size))
+                {
+                    return true;
+                }
+            }
+
+            if (split.Length == 1)
+            {
+                return false;
+            }
+
+            var decimals = split[1].Split(Underscore).Select(s => s.Length).ToArray();
```

Comment:
a [decimal](https://dictionary.cambridge.org/dictionary/english/decimal) is 
> relating to or expressed in a system of counting based on the number ten

please rename to [digit ](https://en.wiktionary.org/wiki/digit#Noun), as it can be in hexa as well

Reply:
also, as above - this if `digitGroupSizes` rather than the digits themselves

Reply:
and again this is a bit weird - why are you splitting based on Undescore here and at line 80 as well?